pkg/kv/kvnemesis/validator.go

```diff
@@ -685,84 +780,144 @@ func (v *validator) checkAtomicCommitted(
+				continue
 			}
-			if !o.Timestamp.IsEmpty() {
+
+			// NB: we allow writeTS to change here, since that will be caught by
+			// validation below anyway, and then we can produce better errors since
+			// read timestamps will be filled in.
+			if writeTS.IsEmpty() {
+				writeTS = o.Timestamp
+			}
+
+			if len(o.EndKey) == 0 { // point write
 				mvccKey := storage.MVCCKey{Key: o.Key, Timestamp: o.Timestamp}
 				if err := batch.Delete(storage.EncodeMVCCKey(mvccKey), nil); err != nil {
 					panic(err)
 				}
+			} else { // ranged write
+				suffix := storage.EncodeMVCCTimestampSuffix(o.Timestamp)
+				if err := batch.RangeKeyUnset(o.Key, o.EndKey, suffix, nil); err != nil {
+					panic(err)
+				}
 			}
 		}
 	}
 
-	// Check if any key that was written twice in the txn had the overwritten
-	// writes materialize in kv. Also fill in all the read timestamps first so
-	// they show up in the failure message.
-	var failure string
+	// Iterate through the observations, building up the snapshot visible at each
+	// point in the atomic unit and filling in the valid read times (validating
+	// them later, in a separate loop, for better errors). We also check that only
+	// the most recent writes materialized (i.e. showed up in MVCC). Check if any
+	// key that was written twice in the txn had the overwritten writes
+	// materialize in kv.
 	for idx, observation := range txnObservations {
 		if failure != `` {
 			break
 		}
 		switch o := observation.(type) {
 		case *observedWrite:
-			var mvccKey storage.MVCCKey
-			if lastWriteIdx := lastWriteIdxByKey[string(o.Key)]; idx == lastWriteIdx {
-				// The last write of a given key in the txn wins and should have made it
-				// to kv.
-				mvccKey = storage.MVCCKey{Key: o.Key, Timestamp: o.Timestamp}
-			} else {
-				if o.Materialized {
-					failure = `committed txn overwritten key had write`
+			// Only the most recent write between overlapping mutations makes it into MVCC.
+			// writeTS was populated above as the unique timestamp at which the writes became
+			// visible. We know the operation had writes (we're looking at one now) and so
+			// this operation has either materialized or is covered by a later one that did,
+			// and so we must have a timestamp here. We defer the failure to the next for
+			// loop, as we will have filled in the read timestamps at that time.
+			if writeTS.IsEmpty() {
+				continue
+			}
+
+			_, isLastWrite := lastWritesByIdx[idx]
+
+			if !isLastWrite && o.Timestamp.IsSet() {
+				failure = `committed txn overwritten key had write`
+				break
+			}
+
+			// Make this write visible (at writeTS, regardless of whether it's the
+			// last write or not, since that's the snapshot at which our operation
+			// wrote). Helpfully, pebble will deal with all the partitioning of range
+			// tombstones that might be necessary under the hood. For example, if we
+			// currently have [a,z) and now are adding a Put(b), we'll automatically
+			// have [a,b); b; [b,z) after.
```

Comment:
This isn't accurate. Point keys and range keys exist separately in Pebble, so range key fragmentation is unaffected by point key writes. If that behavior is necessary, it must be implemented via a `RangeKeyDelete` across the point span.